### PR TITLE
Feature - scapegoat tree

### PR DESCRIPTION
**Pull request by mime8**
This PR introduces new Data Structure - Scapegoat Tree. PR includes Scapegoat Tree class, all related classes and unit tests.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Comments in areas I changed are up to date
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have made corresponding changes to the README.md
- [x] I have made corresponding changes to the DIRECTORY.md

### CODE REVIEW

**Review by siriak — CHANGES_REQUESTED** [review]
Hi! Please fix issues Codacy found and make use of Fluent Assertions in tests as we always do. Thanks!

**Comment by mime8** [comment]
@siriak , hi! Yes, I know. I tried to fix them yesterday. Will continue tomorrow. I'll let you know when I am done.

File: DataStructures.Tests/ScapegoatTree/ScapegoatTreeNodeTests.cs [hunk 1]
```diff
@@ -0,0 +1,190 @@
+using System;
+using DataStructures.ScapegoatTree;
+using NUnit.Framework;
+
+namespace DataStructures.Tests.ScapegoatTree
+{
+    [TestFixture]
+    public class ScapegoatTreeNodeTests
+    {
+
```
Comment (siriak) [inline DataStructures.Tests/ScapegoatTree/ScapegoatTreeNodeTests.cs:10]:
```suggestion
```

File: DataStructures.Tests/ScapegoatTree/ScapegoatTreeNodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        [TestCase("B", "A")]
+        public void LeftSetter_OtherKeyPrecedesLeftKey_AddsChild<TKey>(TKey a, TKey b)
+            where TKey : IComparable
+        {
+            var instance = new Node<TKey>(a);
+            var other = new Node<TKey>(b);
+
+            Assert.DoesNotThrow(() => instance.Left = other);
+        }
+
```
Comment (siriak) [inline DataStructures.Tests/ScapegoatTree/ScapegoatTreeNodeTests.cs:58]:
```suggestion
```

**Review by siriak — APPROVED** [review]
LGTM, thanks!

**Pull request merged**